stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+ *
+ * @param f a functor with signature given above
+ * @param dim dimension of the integral
+ * @param a lower limit of integration as vector
+ * @param b upper limit of integration as vector
+ * @param maxEval maximal number of evaluations
+ * @param reqAbsError absolute error
+ * @param reqRelError relative error as vector
+ * @param pars parameters to be passed to f as a struct
+ * @param val correct value of integral
+ *
+ * @return The value of the n-dimensional integral of \f$f\f$ from \f$a\f$ to
+ \f$b\f$.
+ * @throw std::domain_error no errors will be thrown.
+ */
+
+// hcubature
+template <typename F, typename T_pars, typename T_n, typename T_a, typename T_b,
+          typename T_maxEval, typename T_reqAbsError, typename T_reqRelError>
+return_type_t<T_n, T_a, T_b, T_maxEval, T_reqAbsError, T_reqRelError> hcubature(
+    const F& integrand, const T_pars& pars, const T_n& n, const T_a& a,
+    const T_b& b, const T_maxEval& maxEval, const T_reqAbsError& reqAbsError,
+    const T_reqRelError& reqRelError) {
+  using T_return_type
+      = return_type_t<T_n, T_a, T_b, T_maxEval, T_reqAbsError, T_reqRelError>;
+
+  using T_a_ref = ref_type_t<T_a>;
+  using T_b_ref = ref_type_t<T_b>;
+
+  T_a_ref a_ref = a;
+  T_b_ref b_ref = b;
+
+  scalar_seq_view<T_a_ref> a_vec(a_ref);
+  scalar_seq_view<T_b_ref> b_vec(b_ref);
+
+  internal::one_d out;
+  internal::GenzMalik g;
+
+  if (n == 1) {
+    internal::gauss_kronrod(integrand, a_vec.val(0), b_vec.val(0), out, pars);
+  } else {
+    internal::make_GenzMalik(n, g);
+    internal::integrate_GenzMalik(integrand, g, n, a, b, out, pars);
+  }
+  T_return_type numevals
+      = (n == 1) ? 15 : 1 + 4 * n + 2 * n * (n - 1) + std::pow(2, n);
+  T_return_type evals_per_box = numevals;
+  T_return_type kdiv = out.kdivide;
+  T_return_type err = out.err;
+  T_return_type val = out.result;
+  // convergence test
+  if ((err <= fmax(reqRelError * fabs(val), reqAbsError))
+      || ((maxEval != 0) && (numevals >= maxEval))) {
+    internal::clean_GenzMalik(g);
+    return val;
+  }
+
+  std::multiset<internal::Box> ms;
+  ms.insert(internal::make_box(n, a, b, out));
+
+  while (true) {
+    std::multiset<internal::Box>::iterator it;
+    it = ms.begin();
+    internal::Box box = *it;
+    ms.erase(it);
+    // split along dimension kdiv
+    T_return_type w = (box.b[box.kdiv] - box.a[box.kdiv]) / 2;
+    std::vector<double> ma(box.a);
+
+    ma[box.kdiv] += w;
+    std::vector<double> mb(box.b);
+    mb[box.kdiv] -= w;
+
+    if (n == 1) {
+      internal::gauss_kronrod(integrand, ma[0], box.b[0], out, pars);
+    } else {
+      internal::integrate_GenzMalik(integrand, g, n, ma, box.b, out, pars);
+    }
+    internal::Box box1 = make_box(n, ma, box.b, out);
+    ms.insert(box1);
+
+    if (n == 1) {
+      internal::gauss_kronrod(integrand, box.a[0], mb[0], out, pars);
+    } else {
+      internal::integrate_GenzMalik(integrand, g, n, box.a, mb, out, pars);
+    }
```

Comment:
imo I'd rather have one complete path for n==1 and another path for n > 1. I think having both paths embedded forces you to do this in/out style.

Reply:
You mean first have the case n==1 and after this go to case n>1 instead of all the if elses?